[DataGrid] Add `sortingOrder` to `GridColDef`

Quppa opened the pull request:
Adds `sortingOrder` to `GridColDef`, allowing [`sortingOrder` on `DataGrid`](https://mui.com/components/data-grid/sorting/#sort-order) to be overridden on a per-column basis. An example use case is where a name column would naturally be sorted in ascending order, but downloads in descending order.

m4theushw commented on docs/src/pages/components/data-grid/sorting/OrderSortingPerColumnGrid.tsx:
```diff
@@ -0,0 +1,66 @@
+
+export default function OrderSortingPerColumnGrid() {
+  return (
+    <div style={{ height: 400, width: '100%' }}>
```
Could you use `useDemoData` here, instead of hardcoding the rows and columns? The quantity column could be used instead of the downloads.

Quppa replied:
Updated demo code (not sure how readable this is):
```tsx
<DataGrid
  {...data}
  columns={data.columns.map((column) => ({
    ...column,
    sortingOrder:
      column.field === 'quantity' ? ['desc', 'asc', null] : undefined,
  }))}
  sortingOrder={['asc', 'desc', null]}
/>
```

m4theushw reviewed (commented):
Thanks for your contribution. Could you add a test case?

Why did you remove `ColumnUnsortedIcon`? It was added to satisfy #1076.

Quppa commented:
Thanks for the review - `GridColumnUnsortedIcon` seemed redundant since it just returns `ColumnSortedAscendingIcon` or `ColumnSortedDescendingIcon` based on `nextSortDirection`, and that logic now lives in `GridColumnHeaderSortIcon`, but I see that it's still useful for the reasons described in #1076. That said, there isn't any logic left in `GridColumnUnsortedIcon.tsx` now - is it fine to leave it in this state?

Quppa commented:
Added some tests and switched to passing along `nextSortDirection` using a data attribute on `Icon` (`data-nextsortdirection`). Not at all sure that this is the most elegant approach - am I missing an obvious way to pass a custom attribute (`nextSortDirection`) to the DOM element `Icon`?

m4theushw commented on packages/grid/_modules_/grid/components/columnHeaders/GridColumnHeaderSortIcon.tsx:
```diff
@@ -30,24 +31,35 @@ const useUtilityClasses = (ownerState: OwnerState) => {
+    Icon = icons.ColumnUnsortedIcon;
+    const [nextSortDirection] = sortingOrder;
+    return Icon ? (
+      <Icon fontSize="small" className={className} data-nextsortdirection={nextSortDirection} />
```
Do you need `data-nextsortdirection` for your application? Is not `aria-sort` already sufficient? I'm asking because the logic could be moved to `GridColumnHeaderSortIcon` if the attribute is only used to decide which icon to display.

Quppa replied:
I think this is only used for the mouse-over hint for the next sort direction when the current state is unsorted. I couldn't see an obvious way to contain the logic to `GridColumnHeaderSortIcon` while still returning `GridColumnUnsortedIcon` for the unsorted state.

Good point about `aria-sort` - I've switched to that instead of the custom data attribute.

m4theushw replied:
`aria-sort` can't be used twice. 

From https://developer.mozilla.org/en-US/docs/Web/Accessibility/ARIA/Attributes/aria-sort

> If a grid or table provides sort functionality, the aria-sort attribute should be set to either ascending or descending (or other) on the header cell element for the sorted column or row.

I pushed a small change to pass the `sortingOrder` prop directly to the icon, no data attribute. Are you OK with this approach?

Quppa replied:
Thanks - looks great and is what I'd intended to do in https://github.com/mui-org/material-ui-x/pull/3449/commits/b94f3bf977b2eb2a705bc111cb801f01999c0cab (still confused what the difference between that and your working https://github.com/mui-org/material-ui-x/pull/3449/commits/a3b11f1a793aeda8598077ec14621b4f2a2811c6 version is).

m4theushw replied:
In b94f3bf you were using the `sortingOrder` prop and also spreading it into the DOM element, that's reason for the "React does not recognize the `sortingOrder` prop on a DOM element." warning. In a3b11f1, I don't spread it into the DOM node.

m4theushw commented on packages/grid/x-data-grid/src/tests/sorting.DataGrid.test.tsx:
```diff
@@ -91,6 +91,67 @@ describe('<DataGrid /> - Sorting', () => {
     expect(getColumnValues()).to.deep.equal(['Adidas', 'Puma', 'Nike']);
   });
 
+  it('should only allow ascending sorting using sortingOrder', () => {
```
Thanks for adding tests for the component-wide `sortingOrder`. 💪

Quppa replied:
No problem - I did notice some funky (existing) behaviour when `null` is used in `sortingOrder` in somewhere other than the last position, e.g. `['asc', null, 'desc']` will always cycle between sorting by ascending and the initial order, never descending - I wonder if this should be noted in the docs or even throw an error at runtime.

m4theushw replied:
I guess this prop should only accept `null` at the edges, never in the middle. It might be confusing to go from ascending, no sorting applied, and descending. Let's wait for someone to open an issue with an use case that requires `null` in this position.
